### ISSUE TITLE
[BUG] FlashInfer TRTLLM fused_moe.trtllm_fp8_block_scale_moe: Incorrect Routing with All-Negative Logits

### ISSUE THREAD

**Issue by vadiklyutiy** [flashinfer-ai/flashinfer#2822]
## Summary

`flashinfer.fused_moe.trtllm_fp8_block_scale_moe` (the monolithic variant) produces incorrect expert routing when all router logits are negative. Given byte-identical inputs, the monolithic kernel's output diverges completely (cosine similarity ~0.3) from the modular variant (`trtllm_fp8_block_scale_routed_moe`) which receives pre-computed routing from Python. This causes 0% model accuracy for affected models.

## Background: Monolithic vs Modular

The FlashInfer TRTLLM FP8 MoE backend provides two kernel variants:

- **Monolithic** (`trtllm_fp8_block_scale_moe`): receives raw router logits and performs softmax routing internally inside the CUDA kernel. One fused call handles routing + expert computation.
- **Modular** (`trtllm_fp8_block_scale_routed_moe`): receives pre-computed top-k expert IDs and weights (routing done externally in Python using float32 softmax). The kernel only handles expert computation.

Both variants share the same expert GEMM implementation. The difference is solely in how routing (softmax + top-k selection + renormalization) is performed.

## Bug

When `routing_method_type=1` (Renormalize) and **all router logits are negative**, the monolithic kernel's internal routing selects different experts and assigns different weights compared to the standard Python float32 softmax + top-k. The outputs are uncorrelated.

**Trigger condition**: `max(router_logits) < 0` across all experts for a given token.

**Measured divergence** (single MoE layer, same inputs):

| Logit range | Monolithic vs Modular cosine |
|---|---|
| [-8, +9] (normal) | 1.000 |
| [-14, +3] (some positive) | 1.000 |
| [-18, -1] (all negative) | **0.237** |

## Affected Models

vLLM. **Qwen/Qwen3.5-35B-A3B-FP8** and **Qwen/Qwen3.5-122B-A10B-FP8** — for a tested prompt ("What is the capital of France?", 10 generated tokens), the gate network in the 35B model produces all-negative router logits (range [-11, -1], mean ~-6) across all 40 MoE layers and all generation steps. This triggers the bug in every layer of every forward pass. On the GSM8K benchmark (1319 questions), the monolithic path produces 0% accuracy for both the 35B and 122B models.

The larger **Qwen/Qwen3.5-397B-A17B-FP8** model is not affected because its gate produces some positive logits (max up to +6.88) in later layers during decode, avoiding the trigger condition.

All models use the same gate architecture (`nn.Linear(hidden_size, num_experts, bias=False)`) — the difference is purely in the trained weight values.

## Workaround

Switching from the monolithic to the modular kernel variant restores model accuracy. The modular variant performs routing externally in Python (float32 softmax + top-k), then passes pre-computed expert IDs and weights to the kernel. With this change, Qwen3.5-35B-A3B-FP8 accuracy is restored.

## Reproduction

Requires: NVIDIA Blackwell GPU, flashinfer with TRTLLM MoE support.

```bash
python -m pytest test_trtllm_ep_bug.py -v -s 
```

<details>
<summary>Unit test code (test_trtllm_ep_bug.py)</summary>

```python
"""
Reproduce: trtllm_fp8_block_scale_moe monolithic kernel produces incorrect
routing when all router logits are negative (flat softmax distribution).

The monolithic kernel's internal Renormalize routing diverges from the
modular kernel (Python f32 routing) when all 256 expert logits are negative.
This occurs naturally in Qwen3.5 models where the gate layer produces
all-negative logits.

Trigger: when max(router_logits) < 0, the kernel's internal softmax
routing selects different top-8 experts than Python's float32 softmax,
producing uncorrelated MoE output (cosine ~0.3).
"""
import pytest
import torch


@pytest.fixture
def model_weights():
    """Load real Qwen3.5-35B-A3B-FP8 layer-0 EP0 weights if available,
    otherwise use random weights (bug still reproduces)."""
    from pathlib import Path
    dump = Path(__file__).parent / "kernel_dump_ep0.pt"
    if dump.exists():
        d = torch.load(str(dump), map_location="cpu", weights_only=False)
        return (d["w1"].cuda(), d["w1s"].cuda(),
                d["w2"].cuda(), d["w2s"].cuda())

    torch.manual_seed(0)
    device = "cuda"
    ne, inter, hid = 128, 512, 2048
    w1 = torch.randn(ne, 2*inter, hid, dtype=torch.bfloat16, device=device
                      ).to(torch.float8_e4m3fn)
    w1s = torch.ones(ne, 2*inter//128, hid//128, dtype=torch.float32,
                      device=device)
    w2 = torch.randn(ne, hid, inter, dtype=torch.bfloat16, device=device
                      ).to(torch.float8_e4m3fn)
    w2s = torch.ones(ne, hid//128, inter//128, dtype=torch.float32,
                      device=device)
    return w1, w1s, w2, w2s


def _run_both_kernels(logits, hidden_fp8, scale, w1, w1s, w2, w2s,
                      num_experts=256, local_ne=128, offset=0, top_k=8):
    import flashinfer
    from flashinfer.fused_moe import Fp8QuantizationType

    mono = flashinfer.fused_moe.trtllm_fp8_block_scale_moe(
        routing_logits=logits, routing_bias=None,
        hidden_states=hidden_fp8, hidden_states_scale=scale,
        gemm1_weights=w1, gemm1_weights_scale=w1s,
        gemm2_weights=w2, gemm2_weights_scale=w2s,
        num_experts=num_experts, top_k=top_k,
        n_group=0, topk_group=0,
        intermediate_size=w2.shape[2],
        local_expert_offset=offset, local_num_experts=local_ne,
        routed_scaling_factor=1.0, routing_method_type=1,
        use_shuffled_weight=False,
        fp8_quantization_type=Fp8QuantizationType.DeepSeekFp8,
    )

    scores = torch.softmax(logits.float(), dim=-1)
    tw, ti = torch.topk(scores, k=top_k, dim=-1)
    tw = tw / tw.sum(dim=-1, keepdim=True)
    packed = (ti.int() << 16) | tw.to(torch.bfloat16).view(torch.int16)

    mod = flashinfer.fused_moe.trtllm_fp8_block_scale_routed_moe(
        topk_ids=packed, routing_bias=None,
        hidden_states=hidden_fp8, hidden_states_scale=scale,
        gemm1_weights=w1, gemm1_weights_scale=w1s,
        gemm2_weights=w2, gemm2_weights_scale=w2s,
        num_experts=num_experts, top_k=top_k,
        n_group=None, topk_group=None,
        intermediate_size=w2.shape[2],
        local_expert_offset=offset, local_num_experts=local_ne,
        routed_scaling_factor=None, routing_method_type=1,
        use_shuffled_weight=False, weight_layout=0,
        fp8_quantization_type=Fp8QuantizationType.DeepSeekFp8,
    )

    mf, rf = mono.float(), mod.float()
    if mf.norm() < 1e-6 and rf.norm() < 1e-6:
        return 1.0
    return torch.nn.functional.cosine_similarity(
        mf.reshape(1, -1), rf.reshape(1, -1)).item()


@pytest.mark.parametrize("logit_shift", [0.0, -6.0, -10.0])
def test_all_negative_logits_trigger_routing_divergence(model_weights,
                                                         logit_shift):
    """
    When all router logits are negative (shifted below zero), the monolithic
    kernel's Renormalize routing diverges from Python's float32 routing.

    shift=0:   logits in [-8, +9] -> cosine ~1.0 (OK)
    shift=-6:  logits in [-14, +3] -> cosine ~1.0 (OK, still some positive)
    shift=-10: logits in [-18, -1] -> cosine ~0.3 (BUG: all negative)

    FAILS for shift=-10 because cosine drops below 0.9.
    """
    w1, w1s, w2, w2s = model_weights
    device = "cuda"
    torch.manual_seed(42)

    logits = (torch.randn(4, 256, dtype=torch.bfloat16, device=device) * 3.0
              + logit_shift)
    hidden = torch.randn(4, 2048, dtype=torch.bfloat16, device=device
                         ).to(torch.float8_e4m3fn)
    scale = torch.ones(16, 4, dtype=torch.float32, device=device) * 0.01

    cos = _run_both_kernels(logits, hidden, scale, w1, w1s, w2, w2s)

    assert cos > 0.9, (
        f"Monolithic kernel routing diverges from modular (cosine={cos:.4f}) "
        f"when logits are shifted by {logit_shift} "
        f"(range [{logits.float().min():.1f}, {logits.float().max():.1f}]). "
        f"The kernel's internal Renormalize routing fails with all-negative "
        f"logits, producing incorrect expert selections."
    )
```

</details>

**Comment by vadiklyutiy** [comment]
Hi, 

Is there some news about this issue?

**Comment by yweng0828** [comment]
Hi, thank you for reporting this issue and providing a detailed description and method for reproducing it.

I used the latest ToT for testing, and `model_weights()` used random weights.
```
commit c802a05d9806a3d8ec72014f9cec4d67da0d5e8e (HEAD -> main, tag: nightly-v0.6.12-20260527, origin/main, origin/HEAD)
Author: Wenxuan Tan <wtan45@wisc.edu>
Date:   Tue May 26 16:41:09 2026 -0700
```

I found that I couldn't reproduce the problem. Even after adjusting `logit_shift` to `-1000.0`, it still didn't work. ALL PASS.
I tried directly inputting a random logits, ensuring they were all negative. But still, both kernels produced the same results.

Could you please verify this using the latest code? If you still encounter the problem, could you please provide reproducible input data?

Thanks.

**Comment by aleozlx** [comment]
possibly fix was from #2803
